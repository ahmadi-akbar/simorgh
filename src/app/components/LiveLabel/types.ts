export interface LiveLabelProps {
  offScreenText?: string;
  lang?: string;
  id?: string;
  className?: string;
}

export interface TextProps {
  offScreenText?: string;
  lang?: string;
  id?: string;
}

export interface PulseProps {
  className?: string;
  width: string;
  height: string;
}
