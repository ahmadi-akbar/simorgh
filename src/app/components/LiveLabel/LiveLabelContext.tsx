import React from 'react';

const LiveLabelContext = React.createContext({});
export default LiveLabelContext;
