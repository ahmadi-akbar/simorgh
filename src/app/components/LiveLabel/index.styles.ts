import pixelsToRem from '#app/utilities/pixelsToRem';
import { css, Theme, keyframes } from '@emotion/react';

const pulseAnimation = keyframes({
  '0%': {
    opacity: 1,
  },
  '50%': {
    opacity: 0,
  },
  '100%': {
    opacity: 1,
  },
});

const styles = {
  liveLabelText: ({ spacings, fontVariants }: Theme) =>
    css({
      display: 'inline-block',
      marginInlineEnd: `${spacings.FULL}rem`,
      ...fontVariants.sansBold,
    }),

  pulseContainer: ({ spacings }: Theme) =>
  marginRight: '0.25rem',
      width: `${spacings.DOUBLE}rem`,
      height: `${spacings.DOUBLE}rem`,
      verticalAlign: '-0.125rem',
    }),

  pulsingCircle: () =>
    css({
      '@media (prefers-reduced-motion: no-preference)': {
        animation: `${pulseAnimation} 1.25s cubic-bezier(0.455, 0.03, 0.515, 0.955) -0.4s 3`,
      }),
    
  liveLabelCircle: ({ mq, palette, spacings }: Theme) =>
    css({
      display: 'inline-block',
      width: `${pixelsToRem(15)}rem`,
      height: `${pixelsToRem(15)}rem`,
      background: `radial-gradient(circle, transparent 53%, ${palette.LIVE_DARK} 53.5%, ${palette.LIVE_DARK} 54%)`,
      marginInlineEnd: `${spacings.HALF}rem`,
      position: 'relative',
      transform: 'translate(0, 12%)',
      [mq.GROUP_2_MIN_WIDTH]: {
        width: `${spacings.DOUBLE}rem`,
        height: `${spacings.DOUBLE}rem`,
      },
    }),

  firstPromo: ({ mq }: Theme) =>
    css({
      width: `${pixelsToRem(20)}rem`,
      height: `${pixelsToRem(20)}rem`,
      [mq.GROUP_2_ONLY]: {
        width: `${pixelsToRem(22)}rem`,
        height: `${pixelsToRem(22)}rem`,
      },
      [mq.GROUP_3_MIN_WIDTH]: {
        width: `${pixelsToRem(28)}rem`,
        height: `${pixelsToRem(28)}rem`,
      },
    }),
};

export default styles;
