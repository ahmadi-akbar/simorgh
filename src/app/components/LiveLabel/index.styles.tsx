import { css, Theme } from '@emotion/react';

const styles = {
  liveLabel: ({ palette, fontVariants }: Theme) =>
    css({
      color: palette.POSTBOX,
      display: 'inline-block',
      ...fontVariants.sansBold,
    }),
  textRtl: ({ spacings }: Theme) =>
    css({
      marginLeft: `${spacings.FULL}rem`,
    }),
  textLtr: ({ spacings }: Theme) =>
    css({
      marginRight: `${spacings.FULL}rem`,
    }),
};

export default styles;

// Old CSS below leaving in for reference until work is completed

// const StyledSpan = styled.span`
//   ${({ service }) => getSansBold(service)}
//   color: ${props => props.theme.palette.POSTBOX};
//   display: inline-block;
//   ${({ dir }) =>
//     dir === 'rtl'
//       ? `margin-left: ${GEL_SPACING};`
//       : `margin-right: ${GEL_SPACING};`}
// `;
