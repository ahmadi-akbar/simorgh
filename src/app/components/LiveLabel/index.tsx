/** @jsx jsx */
/** @jsxRuntime classic */
import { PropsWithChildren, useContext } from 'react';
import { jsx } from '@emotion/react';
import VisuallyHiddenText from '../VisuallyHiddenText';
import { ServiceContext } from '../../contexts/ServiceContext';
import styles from './index.styles';
import Text from '../Text';

interface LiveLabelProps {
  offScreenText?: string;
  lang?: string;
  id?: string;
}

const LiveLabel = ({
  lang = 'en-GB',
  id = '',
  children,
  offScreenText,
}: PropsWithChildren<LiveLabelProps>) => {
  const { dir, translations } = useContext(ServiceContext);

  const { liveLabel } = translations.media;

  // As screenreaders mispronounce the word 'LIVE', we use visually hidden
  // text to read 'Live' instead, which screenreaders pronounce correctly.
  const liveLabelIsEnglish = liveLabel === 'LIVE';

  let screenReaderText;
  let ariaHidden = false;

  if (liveLabelIsEnglish) {
    ariaHidden = true;
  }
  if (offScreenText) {
    screenReaderText = `${offScreenText}, `;
  } else if (liveLabelIsEnglish) {
    screenReaderText = 'Live, ';
  } else {
  // Otherwise, the screenreader will pause after reading the word / translation of "Live"  
    screenReaderText = ', ';
  }

  return (
    // lines 27, 56,66, 31 concerning with id are a temporary fix for the a11y nested span's bug experienced in TalkBack, refer to the following issue: https://github.com/bbc/simorgh/issues/9652
    // eslint-disable-next-line jsx-a11y/aria-role
    <Text as="span" id={id} fontVariant="sansBold">
      <span
        css={styles.liveLabel}
        dir={dir}
        {...(ariaHidden && { 'aria-hidden': 'true' })}
      >
        {`${liveLabel} `}
      </span>

      {screenReaderText && (
        <VisuallyHiddenText lang={lang}>
          {`${screenReaderText}`}
        </VisuallyHiddenText>
      )}
      {children}
    </Text>
  );
};

export default LiveLabel;
