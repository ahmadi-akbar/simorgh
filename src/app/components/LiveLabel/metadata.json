{
  "alpha": false,
  "lastUpdated": {
    "day": 18,
    "month": "October",
    "year": 2023
  },
  "uxAccessibilityDoc": {
    "done": false,
    "reference": {
      "url": "",
      "label": "Screen Reader UX"
    }
  },
  "acceptanceCriteria": {
    "done": false,
    "reference": {
      "url": "",
      "label": "Accessibility Acceptance Criteria"
    }
  },
  "swarm": {
    "done": false,
    "reference": {
      "url": "",
      "label": "Accessibility Swarm Notes"
    }
  }
}
