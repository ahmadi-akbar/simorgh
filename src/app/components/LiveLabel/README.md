## Description

The `LiveLabel` component implements a span for use on live content.

## Props

| Argument          | Type    | Required | Default   | Example                 |
| ----------------- | ------- | -------- | --------- | ----------------------- |
| offScreenText     | string  | no       | `null`    | `'Watch Live'`          |
| lang              | string  | no       | `'en-GB'` | `'en-GB'`               |
| children          | node    | no       | `null`    | `<span>Headline</span>` |

## Usage

```jsx
import LiveLabel from '#app/components/LiveLabel';
import { Headline, Link } from '#psammead/psammead-story-promo/src';
import latin from '../../../components/ThemeProvider/fontScripts/latin';

<Headline script={latin} service="news">
  <Link href="https://www.bbc.co.uk/news">
    <LiveLabel offscreenText='Live'>
      The headline of the live promo
    </LiveLabel>
  </Link>
</Headline>;
```

### When to use this component

The `LiveLabel` component can be used to show a promo for a Live page.


### Accessibility notes

The `LiveLabel` example above shows this component being hidden to screen readers with visually hidden text rendered alongside it. This is to ensure the screen reader announces the word 'Live' correctly. This does not need to be accounted for in other languages.
