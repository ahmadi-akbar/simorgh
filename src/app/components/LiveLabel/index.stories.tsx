import React, { useContext } from 'react';
import pathOr from 'ramda/src/pathOr';
import styled from '@emotion/styled';
import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';
import { withServicesKnob } from '../../legacy/psammead/psammead-storybook-helpers/src';
import { Headline, Link } from '../../legacy/psammead/psammead-story-promo/src';
import LiveLabel from '../../legacy/psammead/psammead-live-label/src/index';
import md from '../../legacy/psammead/psammead-live-label/README.md';
import { StoryProps } from '../../models/types/storybook';
import { ThemeProvider } from '../ThemeProvider';
import {
  ServiceContext,
  ServiceContextProvider,
} from '../../contexts/ServiceContext';
import services from '../../../server/utilities/serviceConfigs';
import { ToggleContextProvider } from '../../contexts/ToggleContext';
import { RequestContextProvider } from '../../contexts/RequestContext';
import { ARTICLE_PAGE } from '../../routes/utils/pageTypes';

interface Props extends StoryProps {
  dir: string;
  ariaHidden: boolean;
  offScreenText: string;
  liveText?: string;
}

const Wrapper = styled.div`
  position: relative;
`;

//   .add(
//     'with children',
//     ({ text: headline, script, service, dir }) => (
//       <Wrapper>
//         <Headline script={script} service={service}>
//           <Link href="https://www.bbc.co.uk/news">
//             <LiveLabel
//               service={service}
//               dir={dir}
//               ariaHidden
//               offScreenText="Live"
//             >
//               {headline}
//             </LiveLabel>
//           </Link>
//         </Headline>
//       </Wrapper>
//     ),
//     {
//       notes,
//     },
//   );

const Component = ({
  service,
  variant,
  dir,
  ariaHidden,
  offScreenText = 'Watch Live',
}: Props) => {
  const { liveLabel } = services[service].default.translations.media;
  return (
    <Wrapper>
      <Link href="https://www.bbc.co.uk/news">
        <LiveLabel
          service={service}
          dir={dir}
          ariaHidden
          offScreenText="Live"
          liveText={liveLabel}
        />
      </Link>
    </Wrapper>
  );
};

export default {
  title: 'New Components/Live Label',
  Component,
  decorators: [withKnobs(), withServicesKnob({ defaultService: 'pidgin' })],
  parameters: {
    docs: {
      page: md,
    },
  },
};

export const WithEnglishLiveText = ({ service, variant, dir }: Props) => (
  <Component
    service={service}
    variant={variant}
    dir={dir}
    ariaHidden
    offScreenText="Live"
  />
);

export const WithLocalisedLiveText = ({ service, variant, dir }: Props) => (
  <Component
    service={service}
    variant={variant}
    dir={dir}
    ariaHidden
    offScreenText="Live"
  />
);

export const WithCustomOffscreenText = ({ service, variant, dir }: Props) => (
  <Component
    service={service}
    variant={variant}
    dir={dir}
    ariaHidden
    offScreenText="Watch Live"
  />
);

// export const WithChildren = ({ service, variant, dir }:Props) => (

// )
