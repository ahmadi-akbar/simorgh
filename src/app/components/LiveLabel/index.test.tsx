import React from 'react';
import { render } from '../react-testing-library-with-providers';
import LiveLabel from '.';

describe('LiveLabel', () => {
  it('should render correctly with localised live text', () => {
    const { container } = render(<LiveLabel />, {
      service: 'pidgin',
    });
    expect(container).toMatchSnapshot();
  });
  it('should render correctly with English live text', () => {
    const { container } = render(<LiveLabel />, {
      service: 'russian',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with English live text and children', () => {
    const { container } = render(<LiveLabel> this is a headline </LiveLabel>, {
      service: 'news',
    });
    expect(container).toMatchSnapshot();
  });
  it('should render correctly with custom offscreen text', () => {
    const { container } = render(<LiveLabel offScreenText="Watch Live" />);
    expect(container).toMatchSnapshot();
  });

  it('should correctly render for RTL service', () => {
    const { container } = render(<LiveLabel />, {
      service: 'arabic',
    });
    expect(container).toMatchSnapshot();
  });
});
