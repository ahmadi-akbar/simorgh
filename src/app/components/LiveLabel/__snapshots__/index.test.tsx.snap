// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveLabel should correctly render for RTL service 1`] = `
.emotion-0 {
  color: #141414;
  font-size: 0.9375rem;
  line-height: 1.5rem;
  font-family: "BBC Reith Qalam","Times New Roman",Arial,Verdana,Geneva,Helvetica,sans-serif;
  font-style: normal;
  font-weight: 700;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.5rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.5rem;
  }
}

.emotion-1 {
  color: #B80000;
  display: inline-block;
  -webkit-margin-end: 0.5rem;
  margin-inline-end: 0.5rem;
}

.emotion-2 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

<div>
  <span
    class="emotion-0"
    role="text"
  >
    <span
      aria-hidden="true"
      class="emotion-1"
      dir="rtl"
    >
      LIVE 
    </span>
    <span
      class="emotion-2"
      lang="en-GB"
    >
      Live, 
    </span>
  </span>
</div>
`;

exports[`LiveLabel should render correctly with English live text 1`] = `
.emotion-0 {
  color: #141414;
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: normal;
  font-weight: 700;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

.emotion-1 {
  color: #B80000;
  display: inline-block;
  -webkit-margin-end: 0.5rem;
  margin-inline-end: 0.5rem;
}

.emotion-2 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

<div>
  <span
    class="emotion-0"
    role="text"
  >
    <span
      aria-hidden="true"
      class="emotion-1"
      dir="ltr"
    >
      LIVE 
    </span>
    <span
      class="emotion-2"
      lang="en-GB"
    >
      Live, 
    </span>
  </span>
</div>
`;

exports[`LiveLabel should render correctly with English live text and children 1`] = `
.emotion-0 {
  color: #141414;
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: normal;
  font-weight: 700;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

.emotion-1 {
  color: #B80000;
  display: inline-block;
  -webkit-margin-end: 0.5rem;
  margin-inline-end: 0.5rem;
}

.emotion-2 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

<div>
  <span
    class="emotion-0"
    role="text"
  >
    <span
      aria-hidden="true"
      class="emotion-1"
      dir="ltr"
    >
      LIVE 
    </span>
    <span
      class="emotion-2"
      lang="en-GB"
    >
      Live, 
    </span>
     this is a headline 
  </span>
</div>
`;

exports[`LiveLabel should render correctly with custom offscreen text 1`] = `
.emotion-0 {
  color: #141414;
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: normal;
  font-weight: 700;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

.emotion-1 {
  color: #B80000;
  display: inline-block;
  -webkit-margin-end: 0.5rem;
  margin-inline-end: 0.5rem;
}

.emotion-2 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

<div>
  <span
    class="emotion-0"
    role="text"
  >
    <span
      aria-hidden="true"
      class="emotion-1"
      dir="ltr"
    >
      LIVE 
    </span>
    <span
      class="emotion-2"
      lang="en-GB"
    >
      Watch Live, 
    </span>
  </span>
</div>
`;

exports[`LiveLabel should render correctly with english live text 1`] = `
.emotion-0 {
  color: #141414;
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: normal;
  font-weight: 700;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

.emotion-1 {
  color: #B80000;
  display: inline-block;
  -webkit-margin-end: 0.5rem;
  margin-inline-end: 0.5rem;
}

.emotion-2 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

<div>
  <span
    class="emotion-0"
    role="text"
  >
    <span
      aria-hidden="true"
      class="emotion-1"
      dir="ltr"
    >
      LIVE 
    </span>
    <span
      class="emotion-2"
      lang="en-GB"
    >
      Live, 
    </span>
  </span>
</div>
`;

exports[`LiveLabel should render correctly with localised live text 1`] = `
.emotion-0 {
  color: #141414;
  font-size: 0.9375rem;
  line-height: 1.25rem;
  font-family: Helmet,Freesans,Helvetica,Arial,sans-serif;
  font-style: normal;
  font-weight: 700;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-0 {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

.emotion-1 {
  color: #B80000;
  display: inline-block;
  -webkit-margin-end: 0.5rem;
  margin-inline-end: 0.5rem;
}

<div>
  <span
    class="emotion-0"
    role="text"
  >
    <span
      class="emotion-1"
      dir="ltr"
    >
      AS E DE HAPPEN 
    </span>
  </span>
</div>
`;
